### ISSUE TITLE
Propsal: further refactoring on DataTable/TreeTable column related state

### ISSUE THREAD

**Issue by tandraschko** [primefaces/primefaces#6665]
Currently we have the following state, which is also stored in the MVS:

```
    private Map<String, SortMeta> sortBy;
    private Map<String, FilterMeta> filterBy;
    private String orderedColumnsAsString;
    private Map<String, Boolean> visibleColumns;
    private Map<String, String> resizableColumns;
```
and orderedColumns is still not refactored.

TBH i dont like further refactorings for 100% as our current code is quite nice and stable compared to 8.0 BUT the next version will be 10.0 and i dont like to refactor everything again. And we have at least 3-4 weeks before a RC1.

we could merge and unify many many code if we would create a
```
private Map<String, ColumnState> columnState;
```

which would contain all those info on sorting and filtering, width, ordering, resize etc. etc.
we could also unify the SortBuilder and FilterBuidler and attributes, we would only have something like:

```
<p:dataTable columnsConfigs="..." />
```

with
```
ColumnStateBuilder......sortOrder(...).filterValue(....).width(200).visible(false).build();
```

maybe the same on p:column for a single entry.

@Rapster @christophs78 @melloware WDYT?

**Comment by tandraschko** [comment]
i know there are many other things to refactor but this is probably the most core thing currently we could do.
I would really like to reduce the code here, especially in UITable.

I would also like to re-use DataTableFeatures in TreeTable but this are internal refactorings, this one here is a big public refactoring.

**Comment by melloware** [comment]
I think this would be a good refactoring to help get the two Tables more manageable. I know @Rapster has been in there way more than me though.

**Comment by tandraschko** [comment]
more managable and more unified columnSettings instead of X builder, just 1 for all

**Comment by Rapster** [comment]
I don't really have an opinion on this since, as I already mentionned, I think TreeTable is simply "wrong" as it is. My project integrate this component with which we're not really happy about... And I'll probably start working on a new version of it based on DataTable in the next weeks/months (that's also one of the reasons why I'm trying to clean up stuff as much as I can to have a better start).

But since I didn't start anything yet, and don't know when I'll be able to finalize something. I guess, for now removing duplicates and make code more consistent is always a good thing 😉

**Comment by tandraschko** [comment]
I have some customers and they are quite happy with it actually ;D

nevertheless, this change propsed here does a major refactoring independent of TreeTable, it should just merge column related state to one place and its builder/attributes.

any ideas about the name?
ColumnMeta? Also e.g. LazyDataModel will the pass a Map with ColumnMeta. What about the attr compared to filterBy/sortBy?

**Comment by tandraschko** [comment]
and i will work on a PR for it, not just commit it :P

**Comment by melloware** [comment]
I like `ColumnMeta`.

**Comment by Rapster** [comment]
> and i will work on a PR for it, not just commit it :P

Yes please...

So ColumnMeta will contain everything from FilterMeta/SortMeta?

**Comment by tandraschko** [comment]
exactly
+ isVisible + width + order/displayOrder

**Comment by Rapster** [comment]
Yeah let's see what the PR look like, I fear this new Meta might be bloater code but still sounds interesting, as only one iteration over JSF tree will be necessary to initalize these. Besides `ColumnMeta`, what about `ColumnDescription` _(as they're not really metadata after all)_? Also, I'm trying to find a better name than columnConfig... I'll sleep on it 😄

**Comment by christophs78** [comment]
I´ll keep an eye on your PR.

**Comment by Rapster** [comment]
One downside I see to that, is that methods like that one:

```java
    public boolean isSortingEnabled() {
        return !getSortByAsMap().isEmpty();
    }
```

You'll have to convert into either a loop (and please don't do that) or store a new field "sortingEnabled" in view state

**Comment by tandraschko** [comment]
Thats what i have in mind (not all fields, getter setter, builder etc are created yet):
https://github.com/primefaces/primefaces/pull/6679/files

this state can then be 1:1 reused for MultiViewState and its a single place of truth of the current important state.

**Comment by tandraschko** [comment]
the change would be so big, really unsure
i will first try to reduce code in UITable soon and to reuse DTFeatures for TT

**Comment by melloware** [comment]
Maybe we save this major refactoring for PF10?  Just cleanup and optimize the current code as best we can?

**Comment by tandraschko** [comment]
I think the next release is 10 of you follow Twitter. But yeah, 11 ist fine

**Comment by melloware** [comment]
HA yeah I have not been following the mentioned making it 10 but I have been referring to it as 9.  Should we rename the 9.0 milestone here now to 10?  The pom.xml from 9.0-SNAPSHOT to 10.0-SNAPSHOT?

**Comment by tandraschko** [comment]
Lets wait, primetek will do it when releasing RC1

**Comment by tandraschko** [comment]
fixed by #6689 for now
